### PR TITLE
Introduce a prefetched stream object

### PR DESCRIPTION
**Pull request by Drvi**
An idea; didn't really profile it (yet!). We use a single task for the entire lifetime of the `Object` and we're compatible with `GzipDecompressorStream` to support compression.

### CODE REVIEW

File: src/object.jl [hunk 1]
```diff
@@ -87,3 +87,110 @@ function Base.readbytes!(x::IOObject, dest::AbstractVector{UInt8}, n::Integer=le
...
+
+function _prefetching_task(io)
+    prefetch_size = io.prefetch_size
+    len = io.len
+    object = io.object
+    ppos = 1
+
+    while len > ppos
+        n = min(bytesavailable(io), prefetch_size)
+        buf = PrefetchBuffer(getRange(object, ppos, n), 1, n)
```
[the remark this answers is not in the document]
Comment (quinnj) [reply src/object.jl:106]: Hmmm.....any idea why it's slower? It should be using basically the same machinery underneath; if it's the sizing, then yeah, I'd say we resize the defaults here to whatever works faster.

**Review by quinnj — APPROVED** [review]
This looks awesome @Drvi; thanks. Should we go ahead and remove IOObject while we're here? (or rename PrefetchedDownlowdStream to IOObject?) I'd rather consolidate the 2 instead of having 2 slightly different things.

File: src/object.jl [hunk 2]
```diff
@@ -87,3 +87,171 @@ function Base.readbytes!(x::IOObject, dest::AbstractVector{UInt8}, n::Integer=le
...
+    data::Vector{UInt8}
+    pos::Int
+    len::Int
+end
+Base.bytesavailable(b::PrefetchBuffer) = b.len - b.pos + 1
+
+function _prefetching_task(io)
+    prefetch_size = io.prefetch_size
+    len = io.len
+    ppos = 1
```
Comment (quinnj) [inline src/object.jl:106]:
```suggestion
    ppos = 0
```

File: src/object.jl [hunk 3]
```diff
@@ -87,3 +87,171 @@ function Base.readbytes!(x::IOObject, dest::AbstractVector{UInt8}, n::Integer=le
...
+    download_buffer = Vector{UInt8}(undef, prefetch_size)
+    download_buffer_next = Vector{UInt8}(undef, prefetch_size)
+
+    while len > ppos
+        n = min(len - ppos + 1, prefetch_size)
+        off = 0
+        rngs = Iterators.partition(ppos:ppos+n-1, 2*1024*1024)
```
Comment (quinnj) [inline src/object.jl:113]:
```suggestion
        rngs = Iterators.partition(ppos:ppos+n 2*1024*1024)
```

File: src/object.jl [hunk 4]
```diff
@@ -47,43 +50,199 @@ function Base.unsafe_copyto!(dest::AbstractVector{UInt8}, doff::Integer, src::Ob
...
+
+    try
+        while true
+            (off, rng, download_buffer) = take!(io.download_queue)
+            HTTP.setheader(headers, contentRange(rng))
+            buffer_view = view(download_buffer, off + 1:off + length(rng))
+            response_stream.data = buffer_view
+            response_stream.maxsize = length(buffer_view)
+            seekstart(response_stream)
+            _ = getObject(object.store, url, headers; credentials, response_stream)
```
Comment (Drvi) [inline src/object.jl:122]: Note for this to work properly without allocating, we need to fix IOBuffer handling in HTTP.jl.

Comment (Drvi) [reply src/object.jl:122]: https://github.com/JuliaWeb/HTTP.jl/pull/987

**Review by quinnj — APPROVED** [review]

File: src/object.jl [hunk 5]
```diff
@@ -47,43 +50,199 @@ function Base.unsafe_copyto!(dest::AbstractVector{UInt8}, doff::Integer, src::Ob
...
+            io.cond.ntasks = length(rngs)
+            for rng in rngs
+                put!(io.download_queue, (off, rng .- 1, download_buffer))
+                off += length(rng)
+            end
+            Base.@lock io.cond.cond_wait begin
+                while true
+                    io.cond.ntasks == 0 && break
+                    wait(io.cond.cond_wait)
+                end
```
Comment (NHDaly) [inline src/object.jl:89]: Sorry for the comments all over the place. GitHub is hard.
-----

This confused me at first. the following would be more clear i think:
```suggestion
                while io.cond.ntasks != 0
                    wait(io.cond.cond_wait)
                end
```
or even better:
```suggestion
                while io.cond.ntasks > 0
                    wait(io.cond.cond_wait)
                end
```

File: test/runtests.jl [hunk 6]
```diff
@@ -263,23 +265,151 @@ end
...
-            i += 1000
+        while i < sizeof(multicsv)
+            nb = i + N > length(multicsv) ? length(multicsv) - i : N
+            readbytes!(ioobj, buf, N)
+            @test view(buf, 1:nb) == view(codeunits(multicsv), i:i+nb-1)
+            i += N
         end
     end
 end
+end # @testset "CloudStore.jl"
```
Comment (NHDaly) [inline test/runtests.jl:415]: Can we please add some tests that thoroughly test all the corner cases of the concurrency? I think directly testing what will happen if any of the HTTP calls throw an exception, and testing some corner cases of the possible ordering of the responses.

Probably the best way to do that would be with Mocking. You can see some examples that do this decently well in the raicode protocol tests, in case that's helpful: https://github.com/RelationalAI/raicode/blob/1660f17e12724686f859c2b326ac5e79dfca6773/test/Server/v2-api.jl#L99
